test/distributed/org/apache/cassandra/distributed/impl/Instance.java

```diff
@@ -552,9 +568,66 @@ public int liveMemberCount()
         }).call();
     }
 
-    public int nodetool(String... commandAndArgs)
+    public NodeToolResult nodetoolResult(boolean withNotifications, String... commandAndArgs)
     {
-        return sync(() -> new NodeTool(new InternalNodeProbeFactory()).execute(commandAndArgs)).call();
+        return sync(() -> {
+            DTestNodeTool nodetool = new DTestNodeTool(withNotifications);
+            int rc =  nodetool.execute(commandAndArgs);
+            return new NodeToolResult(commandAndArgs, rc, new ArrayList<>(nodetool.notifications.notifications), nodetool.latestError);
+        }).call();
+    }
+
+    private static class DTestNodeTool extends NodeTool {
+        private final StorageServiceMBean storageProxy;
+        private final CollectingNotificationListener notifications = new CollectingNotificationListener();
+
+        private Throwable latestError;
+
+        DTestNodeTool(boolean withNotifications) {
+            super(new InternalNodeProbeFactory(withNotifications));
+            storageProxy = InternalNodeProbe.create(withNotifications).getStorageService();
```

Comment:
Will `InternalNodeProbe.create(withNotifications)` cause any jmx connection leak?

Reply:
connection on, since JMX shouldn't be on in dtest.  it is possible this object (DTestNodeTool) leaks since `javax.management.NotificationBroadcasterSupport#listenerList` will still hold reference...

Reply:
added close method to remove listener.